distributed/tests/test_spans.py

```diff
@@ -0,0 +1,136 @@
+from __future__ import annotations
+
+from dask import delayed
+
+from distributed import Client, fire_and_forget
+from distributed.spans import span
+from distributed.utils_test import async_poll_for, gen_cluster, inc
+
+
+@gen_cluster(client=True, nthreads=[("", 1)])
+async def test_spans(c, s, a):
+    x = delayed(inc)(1)
+    with span("my workflow"):
+        with span("p1"):
+            y = x + 1
+
+        @span("p2")
+        def f(i):
+            return i + 1
+
+        z = f(y)
+
+    zp = c.persist(z)
+    assert await c.compute(zp) == 4
+
+    ext = s.extensions["spans"]
+
+    def assert_span(d, *expect):
+        span_id = (c.id, *expect)
+        assert s.tasks[d.key].annotations["span"] == span_id
+        assert a.state.tasks[d.key].annotations["span"] == span_id
+        assert ext.spans[span_id].id == span_id
+
+    assert_span(x)
```

Comment:
Is it the case that every task gets a span defined by the client ID?

If so, my sense is that multi-client workloads are rare, and we're probably already able to attribute the client ID.  My inclination would be to drop Client ID from spans, which leaves the common case of no spans unchanged / clean and without annotations, which seems good to me.

Reply:
> Is it the case that every task gets a span defined by the client ID?

Yes, there's always a root span if the user does not explicitly call `span()` on the client workflow, and it's identified by the client ID.

Last week, @fjetter, @hendrikmakait, @ntabris and I reached consensus in this direction during a high-bandwidth design discussion. The conclusion was that adding an annotation to all tasks is cheap and is a small cost to pay to support for multi-tenancy. The overhead becomes even smaller if the user actually uses the `span()` context manager, as this adds just a single element to a tuple.

Reply:
Alternatively, I'd like to propose that we drop the client id from the span annotations.  I suspect that that won't have any bad effects (please correct me if I'm not thinking of something) and will also isolate the effects of the proposed change to only when the new feature is used.  

As an example, I can imagine changes in the future which turn off certain optimizations if *any* annotations are present, just because we don't know how to handle them.  In this case (which seems not unlikely to me) it would have low impact if annotations are rare (as they are today) but would have very high impact if this change goes through.

Reply:
I'm also not necessarily against adding spans by default into all annotations.  Right now I'm mostly asking for the motivation.  What does this help us with?  I don't know of any use common cases where we really care about knowing the client ID.

Reply:
The main point is that we want to support overlapping `span`s to tag independent workloads. This also means that we want to handle workloads from different clients independently.

Setting one default span per client was almost an afterthought to me: When discussing how to pick default spans, we decided to drop any heuristics and simply use static defaults. (This is the important decision.) Using one default span per client felt like a better default on a multi-tenant setup as it would allow users to always differentiate between independent work triggered by different clients (scheduled jobs/users) without any/much perceived cost. I still think having such a feature would be nice, but I personally don't care if that's hidden behind some flag to avoid some cost.

Reply:
Multi-client workloads are exceedingly rare.  I recommend that we don't worry about them until they come up (I'll bet lots of beers that they don't come up).  If they're the only motivation behind adding annotations to all tasks then, to me, the choice is clear that we don't think much about multi-client workloads and also get to not add annotations in the common case.

Reply:
That works for me, spans can still be set explicitly in multi-client environments to handle workloads independently and whether or not we add annotations in the common case is essentially an implementation detail to me.

Reply:
Removed client ID.

Reply:
Thanks @crusaderky .  I appreciate it.